File: app/sun_moon.cpp
#include "sun_moon.h"
#include "config.h"
#include "app_data.h"
#include <Arduino.h>
#include <math.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

static double deg2rad(double d) { return d * M_PI / 180.0; }
static double rad2deg(double r) { return r * 180.0 / M_PI; }

// Julian Day at 0h UT for a calendar date (Gregorian).
static double julianDay(int y, int m, int d) {
  if (m <= 2) { y -= 1; m += 12; }
  int A = y / 100;
  int B = 2 - A + A / 4;
  return floor(365.25 * (y + 4716)) + floor(30.6001 * (m + 1)) + d + B - 1524.5;
}

// NOAA single-pass solar position -> UTC epoch of the sun at `elevDeg`.
time_t sunEventUTC(int y, int m, int d, double elevDeg, bool morning) {
  double JD = julianDay(y, m, d);
  double T  = (JD - 2451545.0) / 36525.0;

  double L0 = fmod(280.46646 + T * (36000.76983 + 0.0003032 * T), 360.0);
  double M  = 357.52911 + T * (35999.05029 - 0.0001537 * T);
  double e  = 0.016708634 - T * (0.000042037 + 0.0000001267 * T);
  double Mr = deg2rad(M);
  double C  = sin(Mr)     * (1.914602 - T * (0.004817 + 0.000014 * T))
            + sin(2 * Mr) * (0.019993 - 0.000101 * T)
            + sin(3 * Mr) *  0.000289;
  double trueLong = L0 + C;
  double omega  = 125.04 - 1934.136 * T;
  double lambda = trueLong - 0.00569 - 0.00478 * sin(deg2rad(omega));
  double eps0   = 23.0 + (26.0 + (21.448 - T * (46.815 + T * (0.00059 - T * 0.001813))) / 60.0) / 60.0;
  double eps    = eps0 + 0.00256 * cos(deg2rad(omega));
  double delta  = rad2deg(asin(sin(deg2rad(eps)) * sin(deg2rad(lambda))));

  double yv = tan(deg2rad(eps / 2.0)); yv *= yv;
  double eqTime = 4.0 * rad2deg(
        yv * sin(2 * deg2rad(L0))
      - 2 * e * sin(Mr)
      + 4 * e * yv * sin(Mr) * cos(2 * deg2rad(L0))
      - 0.5 * yv * yv * sin(4 * deg2rad(L0))
      - 1.25 * e * e * sin(2 * Mr));

  double zenith = 90.0 - elevDeg;
  double cosH = (cos(deg2rad(zenith)) - sin(deg2rad(LATITUDE)) * sin(deg2rad(delta)))
              / (cos(deg2rad(LATITUDE)) * cos(deg2rad(delta)));
  if (cosH > 1.0 || cosH < -1.0) return 0;    // sun never reaches this elevation

  double H = rad2deg(acos(cosH));
  double minutesUTC = 720.0 - 4.0 * (LONGITUDE + (morning ? H : -H)) - eqTime;
  double midnightUTCepoch = (JD - 2440587.5) * 86400.0;
  return (time_t)llround(midnightUTCepoch + minutesUTC * 60.0);
}

const char* moonPhaseName(float p) {
  if (p < 0.03f || p > 0.97f) return "New Moon";
  if (p < 0.22f)              return "Waxing Crescent";
  if (p < 0.28f)              return "First Quarter";
  if (p < 0.47f)              return "Waxing Gibbous";
  if (p < 0.53f)              return "Full Moon";
  if (p < 0.72f)              return "Waning Gibbous";
  if (p < 0.78f)              return "Last Quarter";
  return "Waning Crescent";
}

// Given a local date's tm, return the next day's date via mktime normalisation.
static void nextLocalDate(const struct tm& today, int& y, int& mo, int& d) {
  struct tm t = today;
  t.tm_mday += 1;
  t.tm_hour = 12; t.tm_min = 0; t.tm_sec = 0;   // noon avoids DST edge issues
  time_t tt = mktime(&t);
  struct tm out;
  localtime_r(&tt, &out);
  y = out.tm_year + 1900; mo = out.tm_mon + 1; d = out.tm_mday;
}

void goldenHourStatus(char* buf, size_t n) {
  struct tm t;
  if (!getLocalTime(&t, 10)) { strncpy(buf, "--", n); return; }
  int y = t.tm_year + 1900, mo = t.tm_mon + 1, d = t.tm_mday;
  time_t now = time(nullptr);

  // Golden hour: sun elevation between -4 deg and +6 deg (plan §8).
  time_t mStart = sunEventUTC(y, mo, d, -4.0, true);   // morning window start
  time_t mEnd   = sunEventUTC(y, mo, d,  6.0, true);   // morning window end
  time_t eStart = sunEventUTC(y, mo, d,  6.0, false);  // evening window start
  time_t eEnd   = sunEventUTC(y, mo, d, -4.0, false);  // evening window end

  if (mStart && now >= mStart && now <= mEnd) { snprintf(buf, n, "now (%dm)", (int)((mEnd - now) / 60)); return; }
  if (eStart && now >= eStart && now <= eEnd) { snprintf(buf, n, "now (%dm)", (int)((eEnd - now) / 60)); return; }

  time_t next = 0;
  if (mStart && mStart > now)      next = mStart;
  else if (eStart && eStart > now) next = eStart;
  else {
    int ny, nmo, nd; nextLocalDate(t, ny, nmo, nd);
    next = sunEventUTC(ny, nmo, nd, -4.0, true);
  }
  if (!next) { strncpy(buf, "--", n); return; }

  long secs = next - now;
  snprintf(buf, n, "in %ldh %02ldm", secs / 3600, (secs % 3600) / 60);
}

void sunmoon_recompute() {
  struct tm t;
  if (!getLocalTime(&t, 10)) return;   // need a valid date first
  int y = t.tm_year + 1900, mo = t.tm_mon + 1, d = t.tm_mday;

  g_data.sunriseToday = sunEventUTC(y, mo, d, -0.833, true);
  g_data.sunsetToday  = sunEventUTC(y, mo, d, -0.833, false);

  int ny, nmo, nd; nextLocalDate(t, ny, nmo, nd);
  g_data.sunriseTomorrow = sunEventUTC(ny, nmo, nd, -0.833, true);
  g_data.sunsetTomorrow  = sunEventUTC(ny, nmo, nd, -0.833, false);

  time_t now = time(nullptr);
  g_data.showingNextDay = (now > g_data.sunsetToday);   // roll forward at sunset

  // Moon phase from synodic month since a known new moon (JD 2451550.1).
  double JDnow = 2440587.5 + (double)now / 86400.0;
  double phase = (JDnow - 2451550.1) / 29.530588853;
  phase -= floor(phase);
  g_data.moonPhase = (float)phase;
  g_data.moonIlluminationPct = (float)((1.0 - cos(2 * M_PI * phase)) / 2.0 * 100.0);

  // Print once per boot for offline almanac verification (plan §8).
  static bool printed = false;
  if (!printed) {
    printed = true;
    char a[32], b[32], c[32], e[32];
    strftime(a, sizeof(a), "%Y-%m-%d %H:%M", localtime(&g_data.sunriseToday));
    strftime(b, sizeof(b), "%H:%M",          localtime(&g_data.sunsetToday));
    strftime(c, sizeof(c), "%m-%d %H:%M",    localtime(&g_data.sunriseTomorrow));
    strftime(e, sizeof(e), "%H:%M",          localtime(&g_data.sunsetTomorrow));
    Serial.println("=== sun/moon (verify vs almanac for your coords) ===");
    Serial.printf("today   sunrise %s  sunset %s\n", a, b);
    Serial.printf("tomorrow sunrise %s  sunset %s\n", c, e);
    Serial.printf("moon phase %.3f (%s)  illumination %.0f%%\n",
                  g_data.moonPhase, moonPhaseName(g_data.moonPhase), g_data.moonIlluminationPct);
  }
}

File: Android/app/src/main/java/ca/garionhk/cydclock/time/SunMoon.kt
// ===========================================================================
// SunMoon.kt -- sunrise, sunset, golden hour and moon phase. Port of sun_moon.cpp
// ===========================================================================
// No API. This is why the panel stays correct with the network down, and it is
// worth keeping that way -- the whole scene works offline.
//
// The NOAA single-pass solar position is ported term for term. It is easy to
// "tidy" one of these polynomials and be twenty minutes wrong in December, so
// the constants are left exactly as the firmware has them.
//
// Zero android.* imports, so all of this is testable on the JVM.
package ca.garionhk.cydclock.time

import ca.garionhk.cydclock.core.AppData
import ca.garionhk.cydclock.data.AppSettings
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.util.Locale
import kotlin.math.acos
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.roundToLong
import kotlin.math.sin
import kotlin.math.tan

object SunMoon {

    private fun deg2rad(d: Double) = d * Math.PI / 180.0
    private fun rad2deg(r: Double) = r * 180.0 / Math.PI

    /** Julian Day at 0h UT for a calendar date (Gregorian). */
    fun julianDay(yIn: Int, mIn: Int, d: Int): Double {
        var y = yIn
        var m = mIn
        if (m <= 2) { y -= 1; m += 12 }
        val a = y / 100
        val b = 2 - a + a / 4
        return floor(365.25 * (y + 4716)) + floor(30.6001 * (m + 1)) + d + b - 1524.5
    }

    /**
     * NOAA single-pass solar position -> UTC epoch seconds of the sun reaching
     * [elevDeg], on the morning or evening side.
     *
     * Returns 0 when the sun never reaches that elevation on that date -- polar
     * day and polar night both land here. Callers must treat 0 as "no such
     * event", not as an epoch.
     */
    fun sunEventUTC(
        y: Int, m: Int, d: Int,
        elevDeg: Double,
        morning: Boolean,
        lat: Double, lon: Double,
    ): Long {
        val jd = julianDay(y, m, d)
        val t = (jd - 2451545.0) / 36525.0

        val l0 = (280.46646 + t * (36000.76983 + 0.0003032 * t)).mod(360.0)
        val mAnom = 357.52911 + t * (35999.05029 - 0.0001537 * t)
        val e = 0.016708634 - t * (0.000042037 + 0.0000001267 * t)
        val mr = deg2rad(mAnom)
        val c = sin(mr) * (1.914602 - t * (0.004817 + 0.000014 * t)) +
            sin(2 * mr) * (0.019993 - 0.000101 * t) +
            sin(3 * mr) * 0.000289
        val trueLong = l0 + c
        val omega = 125.04 - 1934.136 * t
        val lambda = trueLong - 0.00569 - 0.00478 * sin(deg2rad(omega))
        val eps0 = 23.0 + (26.0 + (21.448 - t * (46.815 + t * (0.00059 - t * 0.001813))) / 60.0) / 60.0
        val eps = eps0 + 0.00256 * cos(deg2rad(omega))
        val delta = rad2deg(asin(sin(deg2rad(eps)) * sin(deg2rad(lambda))))

        var yv = tan(deg2rad(eps / 2.0)); yv *= yv
        val eqTime = 4.0 * rad2deg(
            yv * sin(2 * deg2rad(l0)) -
                2 * e * sin(mr) +
                4 * e * yv * sin(mr) * cos(2 * deg2rad(l0)) -
                0.5 * yv * yv * sin(4 * deg2rad(l0)) -
                1.25 * e * e * sin(2 * mr)
        )

        val zenith = 90.0 - elevDeg
        val cosH = (cos(deg2rad(zenith)) - sin(deg2rad(lat)) * sin(deg2rad(delta))) /
            (cos(deg2rad(lat)) * cos(deg2rad(delta)))
        if (cosH > 1.0 || cosH < -1.0) return 0    // sun never reaches this elevation

        val h = rad2deg(acos(cosH))
        val minutesUTC = 720.0 - 4.0 * (lon + (if (morning) h else -h)) - eqTime
        val midnightUTCepoch = (jd - 2440587.5) * 86400.0
        return (midnightUTCepoch + minutesUTC * 60.0).roundToLong()
    }

    fun moonPhaseName(p: Float): String = when {
        p < 0.03f || p > 0.97f -> "New Moon"
        p < 0.22f -> "Waxing Crescent"
        p < 0.28f -> "First Quarter"
        p < 0.47f -> "Waxing Gibbous"
        p < 0.53f -> "Full Moon"
        p < 0.72f -> "Waning Gibbous"
        p < 0.78f -> "Last Quarter"
        else -> "Waning Crescent"
    }

    /**
     * Golden hour: sun elevation between -4 and +6 degrees, computed for both the
     * morning and the evening window.
     *
     * The firmware reached tomorrow's date via mktime normalisation at local
     * noon, specifically to dodge DST edges. LocalDate.plusDays has no such
     * hazard, so the noon dance is not needed here -- but the result is the same
     * date, which is the point.
     */
    fun goldenHourStatus(nowEpoch: Long, zone: ZoneId, lat: Double, lon: Double): String {
        val today = Instant.ofEpochSecond(nowEpoch).atZone(zone).toLocalDate()
        val y = today.year; val mo = today.monthValue; val d = today.dayOfMonth

        val mStart = sunEventUTC(y, mo, d, -4.0, true, lat, lon)
        val mEnd = sunEventUTC(y, mo, d, 6.0, true, lat, lon)
        val eStart = sunEventUTC(y, mo, d, 6.0, false, lat, lon)
        val eEnd = sunEventUTC(y, mo, d, -4.0, false, lat, lon)

        if (mStart != 0L && nowEpoch >= mStart && nowEpoch <= mEnd) {
            return String.format(Locale.US, "now (%dm)", (mEnd - nowEpoch) / 60)
        }
        if (eStart != 0L && nowEpoch >= eStart && nowEpoch <= eEnd) {
            return String.format(Locale.US, "now (%dm)", (eEnd - nowEpoch) / 60)
        }

        var next = 0L
        if (mStart != 0L && mStart > nowEpoch) next = mStart
        else if (eStart != 0L && eStart > nowEpoch) next = eStart
        else {
            val t = today.plusDays(1)
            next = sunEventUTC(t.year, t.monthValue, t.dayOfMonth, -4.0, true, lat, lon)
        }
        if (next == 0L) return "--"

        val secs = next - nowEpoch
        return String.format(Locale.US, "in %dh %02dm", secs / 3600, (secs % 3600) / 60)
    }

    /**
     * Recompute the sun/moon block. Cheap and offline, so it is safe to call on
     * every scene entry -- but the tick loop only calls it on start, on a date
     * change, and when the location or timezone changes.
     */
    fun recompute(d: AppData, s: AppSettings, zone: ZoneId, nowEpoch: Long): AppData {
        val lat = s.latitude
        val lon = s.longitude
        val today: LocalDate = Instant.ofEpochSecond(nowEpoch).atZone(zone).toLocalDate()
        val tomorrow = today.plusDays(1)

        val sunriseToday = sunEventUTC(today.year, today.monthValue, today.dayOfMonth, -0.833, true, lat, lon)
        val sunsetToday = sunEventUTC(today.year, today.monthValue, today.dayOfMonth, -0.833, false, lat, lon)
        val sunriseTomorrow = sunEventUTC(tomorrow.year, tomorrow.monthValue, tomorrow.dayOfMonth, -0.833, true, lat, lon)
        val sunsetTomorrow = sunEventUTC(tomorrow.year, tomorrow.monthValue, tomorrow.dayOfMonth, -0.833, false, lat, lon)

        // DIVERGENCE (fixed): the firmware wrote `showingNextDay = now > sunsetToday`.
        // sunEventUTC returns 0 during polar day and polar night, so above the
        // Arctic circle that reduces to `now > 0` and the scene claims TOMORROW
        // permanently. Guarding on a real sunset costs nothing and is correct
        // everywhere the original was correct.
        val showingNextDay = sunsetToday != 0L && nowEpoch > sunsetToday

        // Moon phase from the synodic month since a known new moon (JD 2451550.1).
        val jdNow = 2440587.5 + nowEpoch / 86400.0
        var phase = (jdNow - 2451550.1) / 29.530588853
        phase -= floor(phase)

        return d.copy(
            sunriseToday = sunriseToday,
            sunsetToday = sunsetToday,
            sunriseTomorrow = sunriseTomorrow,
            sunsetTomorrow = sunsetTomorrow,
            showingNextDay = showingNextDay,
            moonPhase = phase.toFloat(),
            moonIlluminationPct = ((1.0 - cos(2 * Math.PI * phase)) / 2.0 * 100.0).toFloat(),
            goldenHour = goldenHourStatus(nowEpoch, zone, lat, lon),
        )
    }
}
